src/lib/update-lock.ts

```diff
@@ -82,6 +82,75 @@ async function dispose(
 	}
 }
 
+/**
+ * Given a lockfile path `p`, return a tuple [appId, serviceName] of that path.
+ * Paths are assumed to end in the format /:appId/:serviceName/(resin-)updates.lock.
+ */
+function getIdentifiersFromPath(p: string) {
+	const parts = p.split('/');
+	if (parts.pop()?.match(/updates\.lock/) === null) {
+		return [];
+	}
+	const serviceName = parts.pop();
+	const appId = parts.pop();
+	return [appId, serviceName];
+}
+
+type LockedEntity = { appId: number; services: string[] };
+
+/**
+ * A map of locked services by appId.
+ * Exported for tests only; getServicesLockedByAppId is the public generator interface.
+ */
+export class LocksTakenMap extends Map<number, Set<string>> {
+	constructor(lockedEntities: LockedEntity[] = []) {
+		// Construct a Map<number, Set<string>> from user-friendly input args
+		super(
+			lockedEntities.map(({ appId, services }) => [appId, new Set(services)]),
+		);
+	}
+
+	// Add one or more locked services to an appId
+	public add(appId: number, services: string | string[]): void {
+		if (typeof services === 'string') {
+			services = [services];
+		}
+		if (this.has(appId)) {
+			const lockedSvcs = this.get(appId)!;
+			services.forEach((s) => lockedSvcs.add(s));
+		} else {
+			this.set(appId, new Set(services));
+		}
+	}
+
+	// Return an array copy of locked services under an appId
+	public getServices(appId: number): string[] {
+		return this.has(appId) ? Array.from(this.get(appId)!) : [];
+	}
```

Comment:
Since you are extending `Map` I think is best to extend `get` here, however you should be careful to return something the user can modify

```ts
public getServices(appId: number): string[] {
   let svcs = [];
   if (this.has(appId)) {
       svcs = super.get(appId);
   }
   else {
      super.set(appId, svcs);
   }
   return svcs; 
}
```

That way modifications to the returned array will also modify the reference in the Map, same as with https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get

Reply:
I wanted to avoid direct modification and have `getServices` act as a readonly method, hence why I avoided the `get` extension here. For the purposes for which LocksTakenMap is used, there is no use case for a `get` which returns a mutable reference.

Reply:
That is not obvious from the class definition. Some future supervisor developer can still use the other `Map` methods to do things "the wrong way". What you could do is make the `get`method private in that case, to prevent usage.

Reply:
There's no way to extend a Map but modify the types of the interfaces of `get`, so I did the next best thing which was to set the `@private` annotation with a comment on `get`. [Ref](https://github.com/balena-os/balena-supervisor/pull/2234/commits/24d7a2fb63fb9a93e215e3615e5bd8b98279f107#diff-376068891dd599a79a4b6bf84d850688e14f1d5ebaf089f88518f08e40fc6f4bR126-R132)

Reply:
Oh, I thought it was possible. I guess you could also create a new interface that operates on an internal map and not extend `Map`. I think that might result in cleaner code, but I leave it up to you